src/Riok.Mapperly/Descriptors/MemberPathCandidateBuilder.cs

```diff
@@ -1,66 +1,273 @@
 
-        var indices = GetPascalCaseSplitIndices(name).Take(MaxPascalCaseIndices).ToArray();
-        if (indices.Length == 0)
+        if (indicesCount == 0)
+        {
             yield break;
+        }
 
-        // try all permutations, skipping the first because the full string is already yielded
-        var permutationsCount = 1 << indices.Length;
+        var permutationsCount = 1 << indicesCount;
+
+        // Start at 1. Mask 0 means "keep all underscores", which equals the original string.
+        // The original string was already yielded by the main method.
         for (var i = 1; i < permutationsCount; i++)
         {
-            yield return new StringMemberPath(BuildPermutationParts(name, indices, i));
+            yield return new StringMemberPath(BuildPermutationList(name, indices, indicesCount, i, skipDelimiter: true));
+        }
+    }
+
+    /// <summary>
+    /// Simple check to avoid string allocation in ToSnakeCase if not needed
+    /// </summary>
+    private static bool IsSnakeCase(string str)
+    {
+        foreach (var t in str)
+        {
+            if (char.IsUpper(t))
+                return false;
+        }
+
+        return true;
+    }
+
+    private static bool IsUpperSnakeCase(string str)
+    {
+        foreach (var t in str)
+        {
+            if (char.IsLower(t))
+                return false;
+        }
+
+        return true;
+    }
+
+    private static int FillPascalSplitIndices(string str, int[] buffer)
+    {
+        var count = 0;
+        // Start at 1. Splitting at index 0 is invalid for member paths
+        for (var i = 1; i < str.Length && count < buffer.Length; i++)
+        {
+            if (char.IsUpper(str[i]))
+            {
+                buffer[count++] = i;
+            }
         }
+
+        return count;
     }
 
-    private static IEnumerable<string> BuildPermutationParts(string source, int[] splitIndices, int enabledSplitPositions)
+    private static int FillCharIndices(string str, char c, int[] buffer)
     {
-        var lastSplitIndex = 0;
-        var currentSplitPosition = 1;
-        foreach (var splitIndex in splitIndices)
+        var count = 0;
+        for (var i = 0; i < str.Length && count < buffer.Length; i++)
         {
-            if ((enabledSplitPositions & currentSplitPosition) == currentSplitPosition)
+            if (str[i] == c)
             {
-                yield return source.Substring(lastSplitIndex, splitIndex - lastSplitIndex);
-                lastSplitIndex = splitIndex;
+                buffer[count++] = i;
             }
+        }
+
+        return count;
+    }
+
+    private static List<string> BuildPermutationList(string source, int[] indices, int indicesCount, int mask, bool skipDelimiter)
+    {
+        // Pre-calculate exact capacity to avoid internal List resizing
+        // 1 base segment + 1 for every split
+        var capacity = 1 + CountSetBits(mask);
+        var list = new List<string>(capacity);
```

Comment:
It looks like you attempted to optimize for performance. Have you benchmarked this yet?
From my perspective the previous code was easier to follow, and I might be biased, so I would still like to double check whether the added complexity is justified.
There is probably still room for significant optimization by using (ReadOnly)Spans, stackalloc and/or array pools. I'm not sure whether we should move these performance optimizations into a separate PR.